codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java

```diff
@@ -402,20 +392,20 @@ private void verifyUnknownFrame() throws Http2Exception {
         verifyNotProcessingHeaders();
     }
 
-    private void readDataFrame(ChannelHandlerContext ctx, ByteBuf payload, int payloadEndIndex,
+    private void readDataFrame(ChannelHandlerContext ctx, ByteBuf payload,
             Http2FrameListener listener) throws Http2Exception {
         int padding = readPadding(payload);
         verifyPadding(padding);
 
         // Determine how much data there is to read by removing the trailing
         // padding.
-        int dataLength = lengthWithoutTrailingPadding(payloadEndIndex - payload.readerIndex(), padding);
+        int dataLength = lengthWithoutTrailingPadding(payloadLength - payload.readerIndex(), padding);
 
         ByteBuf data = payload.readSlice(dataLength);
```

Comment:
Do you think we could just adjust the writer index on `payload`?

Reply:
So instead of slicing the data buffer just move the writer index of the payload to dataLength?

Reply:
Yes: I think we sliced the payload previously because we didn't want to leak the parent buffer but since we've already sliced it there shouldn't be any harm in sharing that. At a minimum, I expect padding to be rare so if there isn't any padding to slice out we can certainly just send `payload` along.

And to be more precise, I think we should do `payload.writeIndex(payload.readerIndex() + dataLength)`. `payload.readerIndex()` should always be 0 now, but it will make the code more resilient to future changes.

Reply:
Ok cool I'll make that change and saw the payload being further sliced while reading a Go Away frame so I'm updating that as well